File: csharp/Solitaire/KlondikeTests/Move/StockPileRecycleMoveTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Silnith.Game.Deck;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Silnith.Game.Klondike.Move.Tests
{
    [TestClass]
    public class StockPileRecycleMoveTests
    {
        private readonly IReadOnlyList<Card> EmptyListOfCards = Array.Empty<Card>();
        private readonly IReadOnlyDictionary<Suit, IReadOnlyList<Card>> EmptyFoundation = new Dictionary<Suit, IReadOnlyList<Card>>()
        {
            { Suit.Club, Array.Empty<Card>() },
            { Suit.Diamond, Array.Empty<Card>() },
            { Suit.Heart, Array.Empty<Card>() },
            { Suit.Spade, Array.Empty<Card>() },
        };
        private readonly IReadOnlyList<Column> EmptyColumns = new List<Column>()
        {
            new Column(null, null),
            new Column(null, null),
            new Column(null, null),
            new Column(null, null),
            new Column(null, null),
            new Column(null, null),
            new Column(null, null),
        };

        #region FindMoves

        [TestMethod]
        public void TestFindMovesEmptyStockPile()
        {
            Board board = new(EmptyColumns, EmptyListOfCards, 0, EmptyFoundation);

            IEnumerable<StockPileRecycleMove> actual = StockPileRecycleMove.FindMoves(board);

            Assert.IsFalse(actual.Any());
        }

        [TestMethod]
        public void TestFindMovesStockPileAtBeginning()
        {
            List<Card> stockPile = new()
            {
                new Card(Value.Four, Suit.Club),
                new Card(Value.Seven, Suit.Heart),
                new Card(Value.Six, Suit.Heart),
                new Card(Value.Three, Suit.Diamond),
            };
            Board board = new(EmptyColumns, stockPile, 0, EmptyFoundation);

            IEnumerable<StockPileRecycleMove> actual = StockPileRecycleMove.FindMoves(board);

            Assert.IsFalse(actual.Any());
        }

        [TestMethod]
        public void TestFindMovesStockPileInMiddle()
        {
            List<Card> stockPile = new()
            {
                new Card(Value.Four, Suit.Club),
                new Card(Value.Seven, Suit.Heart),
                new Card(Value.Six, Suit.Heart),
                new Card(Value.Three, Suit.Diamond),
            };
            Board board = new(EmptyColumns, stockPile, 2, EmptyFoundation);

            IEnumerable<StockPileRecycleMove> actual = StockPileRecycleMove.FindMoves(board);

            Assert.IsFalse(actual.Any());
        }

        [TestMethod]
        public void TestFindMovesStockPileAtEnd()
        {
            List<Card> stockPile = new()
            {
                new Card(Value.Four, Suit.Club),
                new Card(Value.Seven, Suit.Heart),
                new Card(Value.Six, Suit.Heart),
                new Card(Value.Three, Suit.Diamond),
            };
            Board board = new(EmptyColumns, stockPile, 4, EmptyFoundation);

            IEnumerable<StockPileRecycleMove> actual = StockPileRecycleMove.FindMoves(board);

            IEnumerable<StockPileRecycleMove> expected = new List<StockPileRecycleMove>(1)
            {
                new StockPileRecycleMove(4),
            };
            CollectionAssert.AreEquivalent(expected.ToList(), actual.ToList());
        }

        #endregion

        [TestMethod]
        public void TestSourceIndex()
        {
            StockPileRecycleMove move = new(5);

            Assert.AreEqual(5, move.SourceIndex);
        }

        [TestMethod]
        public void TestHasCards()
        {
            StockPileRecycleMove move = new(5);

            Assert.IsFalse(move.HasCards);
        }

        #region Apply

        [TestMethod]
        public void TestApply()
        {
            List<Column> columns = new()
            {
                new Column(null, null),
                new Column(null, null),
                new Column(null, null),
                new Column(null, null),
                new Column(
                    null,
                    new List<Card>()
                    {
                        new Card(Value.King, Suit.Spade),
                        new Card(Value.Queen, Suit.Heart),
                    }),
                new Column(null, null),
                new Column(null, null),
            };
            List<Card> stockPile = new()
            {
                new Card(Value.Three, Suit.Club),
                new Card(Value.Jack, Suit.Spade),
            };
            int stockPileIndex = stockPile.Count;
            Dictionary<Suit, IReadOnlyList<Card>> foundation = new()
            {
                {
                    Suit.Club,
                    new List<Card>()
                    {
                        new Card(Value.Ace, Suit.Club),
                        new Card(Value.Two, Suit.Club),
                    }
                },
                {
                    Suit.Diamond,
                    Array.Empty<Card>()
                },
                {
                    Suit.Heart,
                    Array.Empty<Card>()
                },
                {
                    Suit.Spade,
                    Array.Empty<Card>()
                },
            };
            Board board = new(columns, stockPile, stockPileIndex, foundation);

            StockPileRecycleMove move = new(stockPileIndex);

            Board actual = move.Apply(board);

            Board expected = new(columns, stockPile, 0, foundation);

            Assert.AreEqual(expected, actual);
        }

        // TODO: Add tests that illegal recycles throw exceptions.

        #endregion

        #region Equals & GetHashCode

        [TestMethod]
        public void TestEquals()
        {
            StockPileRecycleMove move1 = new(5);
            StockPileRecycleMove move2 = new(5);

            Assert.AreEqual(move1, move2);
        }

        [TestMethod]
        public void TestGetHashCode()
        {
            StockPileRecycleMove move1 = new(5);
            StockPileRecycleMove move2 = new(5);

            Assert.AreEqual(move1.GetHashCode(), move2.GetHashCode());
        }

        [TestMethod]
        public void TestEqualsDifferentIndex()
        {
            StockPileRecycleMove move1 = new(5);
            StockPileRecycleMove move2 = new(4);

            Assert.AreNotEqual(move1, move2);
        }

        #endregion

    }
}

File: cpp/Solitaire/KlondikeTests/StockPileRecycleMoveTests.cpp
#include "CppUnitTest.h"

#include "BoardToString.h"

#include <silnith/game/solitaire/move/StockPileRecycleMove.h>

#include <memory>
#include <set>

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

using namespace silnith::game::deck;
using namespace silnith::game::solitaire;
using namespace silnith::game::solitaire::move;

using namespace std;
using namespace std::literals::string_literals;

struct StockPileRecycleMoveComparator
{
public:
	bool operator()(shared_ptr<StockPileRecycleMove> const& lhs, shared_ptr<StockPileRecycleMove> const& rhs) const
	{
		return lhs->get_source_index() < rhs->get_source_index();
	}
};

using StockPileRecycleMoveSet = set<shared_ptr<StockPileRecycleMove>, StockPileRecycleMoveComparator>;

bool operator==(StockPileRecycleMoveSet const& lhs, StockPileRecycleMoveSet const& rhs)
{
	vector<StockPileRecycleMove> lhs_moves{};
	vector<StockPileRecycleMove> rhs_moves{};
	for (shared_ptr<StockPileRecycleMove> ptr : lhs)
	{
		lhs_moves.emplace_back(*ptr);
	}
	for (shared_ptr<StockPileRecycleMove> ptr : rhs)
	{
		rhs_moves.emplace_back(*ptr);
	}
	return lhs_moves == rhs_moves;
}

StockPileRecycleMoveSet ToStockPileRecycleMoveSet(vector<shared_ptr<solitaire_move>> const& vec)
{
	vector<shared_ptr<StockPileRecycleMove>> typed_vec{};
	for (shared_ptr<solitaire_move> move : vec)
	{
		typed_vec.emplace_back(dynamic_pointer_cast<StockPileRecycleMove>(move));
	}
	StockPileRecycleMoveSet s{ typed_vec.begin(), typed_vec.end() };
	return s;
}

template<>
wstring Microsoft::VisualStudio::CppUnitTestFramework::ToString<>(StockPileRecycleMove const& move)
{
	return to_wstring(move);
}

template<>
wstring Microsoft::VisualStudio::CppUnitTestFramework::ToString<>(StockPileRecycleMoveSet const& moves)
{
	wostringstream out{};
	out << L"{"s;
	if (!moves.empty())
	{
		StockPileRecycleMoveSet::const_iterator citer{ moves.cbegin() };
		StockPileRecycleMoveSet::const_iterator cend{ moves.cend() };
		shared_ptr<StockPileRecycleMove> ptr{ *citer };
		out << *ptr;
		citer++;
		while (citer != cend)
		{
			out << L", "s;
			shared_ptr<StockPileRecycleMove> ptr{ *citer };
			out << *ptr;
			citer++;
		}
	}
	out << L"}"s;
	return out.str();
}

namespace SolitaireMoveTests
{
	TEST_CLASS(StockPileRecycleMoveTests)
	{
	private:
		vector<card> empty_list_of_cards{};
		vector<column> empty_columns{
			column{ empty_list_of_cards, empty_list_of_cards },
			column{ empty_list_of_cards, empty_list_of_cards },
			column{ empty_list_of_cards, empty_list_of_cards },
			column{ empty_list_of_cards, empty_list_of_cards },
			column{ empty_list_of_cards, empty_list_of_cards },
			column{ empty_list_of_cards, empty_list_of_cards },
			column{ empty_list_of_cards, empty_list_of_cards },
		};
		map<suit, vector<card>> empty_foundation{
			{ suit::club, empty_list_of_cards },
			{ suit::diamond, empty_list_of_cards },
			{ suit::heart, empty_list_of_cards },
			{ suit::spade, empty_list_of_cards },
		};

	public:
		TEST_METHOD(TestFindMovesEmptyStockPile)
		{
			board new_board{ empty_columns, empty_list_of_cards, 0, empty_foundation };

			vector<shared_ptr<solitaire_move>> moves{ StockPileRecycleMove::find_moves(new_board) };
			StockPileRecycleMoveSet actual{ ToStockPileRecycleMoveSet(moves) };

			StockPileRecycleMoveSet expected{};
			Assert::AreEqual(expected, actual);
		}

		TEST_METHOD(TestFindMovesStockPileAtBeginning)
		{
			vector<card> stock_pile{
				card{ value::four, suit::club },
				card{ value::seven, suit::heart },
				card{ value::six, suit::heart },
				card{ value::three, suit::diamond },
			};
			board new_board{ empty_columns, stock_pile, 0, empty_foundation };

			vector<shared_ptr<solitaire_move>> moves{ StockPileRecycleMove::find_moves(new_board) };
			StockPileRecycleMoveSet actual{ ToStockPileRecycleMoveSet(moves) };

			StockPileRecycleMoveSet expected{};
			Assert::AreEqual(expected, actual);
		}

		TEST_METHOD(TestFindMovesStockPileInMiddle)
		{
			vector<card> stock_pile{
				card{ value::four, suit::club },
				card{ value::seven, suit::heart },
				card{ value::six, suit::heart },
				card{ value::three, suit::diamond },
			};
			board new_board{ empty_columns, stock_pile, 2, empty_foundation };

			vector<shared_ptr<solitaire_move>> moves{ StockPileRecycleMove::find_moves(new_board) };
			StockPileRecycleMoveSet actual{ ToStockPileRecycleMoveSet(moves) };

			StockPileRecycleMoveSet expected{};
			Assert::AreEqual(expected, actual);
		}

		TEST_METHOD(TestFindMovesStockPileAtEnd)
		{
			vector<card> stock_pile{
				card{ value::four, suit::club },
				card{ value::seven, suit::heart },
				card{ value::six, suit::heart },
				card{ value::three, suit::diamond },
			};
			board new_board{ empty_columns, stock_pile, 4, empty_foundation };

			vector<shared_ptr<solitaire_move>> moves{ StockPileRecycleMove::find_moves(new_board) };
			StockPileRecycleMoveSet actual{ ToStockPileRecycleMoveSet(moves) };

			StockPileRecycleMoveSet expected{
				make_shared<StockPileRecycleMove>(4),
			};
			Assert::AreEqual(expected, actual);
		}

		TEST_METHOD(TestGetSourceIndex)
		{
			StockPileRecycleMove move{ 5 };

			Assert::AreEqual(size_t{ 5 }, move.get_source_index());
		}

		TEST_METHOD(TestHasCards)
		{
			StockPileRecycleMove move{ 5 };

			Assert::IsFalse(move.has_cards());
		}

		TEST_METHOD(TestApply)
		{
			vector<column> columns{
				column{
					empty_list_of_cards,
					empty_list_of_cards
				},
				column{
					empty_list_of_cards,
					empty_list_of_cards
				},
				column{
					empty_list_of_cards,
					empty_list_of_cards
				},
				column{
					empty_list_of_cards,
					empty_list_of_cards
				},
				column{
					empty_list_of_cards,
					vector<card>{
						card{ value::king, suit::spade },
						card{ value::queen, suit::heart },
					}
				},
				column{
					empty_list_of_cards,
					empty_list_of_cards
				},
				column{
					empty_list_of_cards,
					empty_list_of_cards
				},
			};
			vector<card> stock_pile{
				card{ value::three, suit::club },
				card{ value::jack, suit::spade },
			};
			size_t stock_pile_index{ stock_pile.size() };
			map<suit, vector<card>> foundation{
				{
					suit::club,
					vector<card>{
						card{ value::ace, suit::club },
						card{ value::two, suit::club },
					}
				},
				{
					suit::diamond,
					empty_list_of_cards
				},
				{
					suit::heart,
					empty_list_of_cards
				},
				{
					suit::spade,
					empty_list_of_cards
				},
			};

			shared_ptr<board> board_ptr{ make_shared<board>(columns, stock_pile, stock_pile_index, foundation) };

			StockPileRecycleMove move{ stock_pile_index };

			shared_ptr<board> actual{ move.apply(board_ptr) };

			board expected{ columns, stock_pile, 0, foundation };
			Assert::AreEqual(expected, *actual);
		}

		TEST_METHOD(TestEquals)
		{
			StockPileRecycleMove move1{ 5 };
			StockPileRecycleMove move2{ 5 };

			Assert::AreEqual(move1, move2);
		}

		TEST_METHOD(TestEqualsDifferentIndex)
		{
			StockPileRecycleMove move1{ 5 };
			StockPileRecycleMove move2{ 4 };

			Assert::AreNotEqual(move1, move2);
		}
	};
}
